### PR TITLE
Add watchdog functionality

### LINKED ISSUE #298: Add option for watchdog service
Apparently the Pi has a hardware watchdog timer, see http://blog.ricardoarturocabral.com/2013/01/auto-reboot-hung-raspberry-pi-using-on.html

It could be a nice feature to add.

### PR DESCRIPTION
**Pull request by goranche**
Fixes #298

### CODE REVIEW

**Comment by Mausy5043** [comment]
Having had some bad experiences with a watchdog ending up rebooting my device in a loop. I'd love the settings to be a bit more relaxed. Would you consider setting the timerdefault at one or two minutes?
This gives the user time to try and fix things via a terminal.

**Comment by goranche** [comment]
> setting the timerdefault at one or two minutes

don't think that should be needed, no... systemd should take care of making sure the hardware watchdog doesn't restart the system... as it's configured in this PR, it will "ping" the watchdog every 10 seconds, so it shouldn't be an issue, except if it fails to do so, in which case you'd be out of luck anyway...

**Comment by Mausy5043** [comment]
I've tested this and it seems to work (module is loaded and setting is inserted correctly). Anybody have a "simple" way of blocking out the kernel for >20s to force the watchdog to cause a reboot? 

I was trying to `git clone https://github.com/raspberrypi/linux.git` and that triggered a reboot. Not sure if it was related to the watchdog though. Did see some `oom-killer`s as it progressed...

**Comment by goranche** [comment]
> blocking out the kernel

not so much the kernel as systemd pinging of the watchdog device...

**Comment by Mausy5043** [comment]
Whatever, as long as it achieves the goal.

Device is up:
```
$ la /dev/w*
crw------- 1 root root 10, 130 Mar 20 18:17 /dev/watchdog
```

**Comment by goranche** [comment]
> Whatever

:smirk: just thought you'd like to understand how the watchdog actually functions...

anyway... not sure what you're trying to test... you can disable systemds watchdog handling, and then do a simple `cat > /dev/watchdog`... as long as you keep sending characters (don't use `V` :thought_balloon: ) into the device, the hardware watchdog timer is reset... when you stop doing that, the device will sooner or later (about 10 seconds in my case) reboot...

**Comment by goranche** [comment]
> don't use V

never mind... it seems the `bcm2708_wdog` doesn't support the "Magic Close" functionality... :smile: 
if you have time to spare, here's some [light reading](https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/Documentation/watchdog/watchdog-api.txt)...

**Comment by Mausy5043** [comment]
I kinda know how it works. Was just wondering if there's a way to starve the watchdog.

**Comment by goranche** [comment]
> I was trying to `git clone ...` and that triggered a reboot

I'm a bit confused / worried by this one, though... :thought_balloon:

**Comment by Mausy5043** [comment]
Yes, I need to do some extra testing to make sure it's unrelated.

**Comment by Mausy5043** [comment]
This log-entry bothers me:

```
Mar 21 16:17:02 rbian systemd[1]: Started udev Coldplug all Devices.
Mar 21 16:17:02 rbian kernel: bcm2708 watchdog, heartbeat=10 sec (nowayout=0)
Mar 21 16:17:02 rbian systemd[1]: Hardware watchdog 'BCM2708', version 1
Mar 21 16:17:02 rbian systemd[1]: Failed to set timeout to 20s: Invalid argument
```

**Comment by Mausy5043** [comment]
This causes crashes
```
$ mkdir tmp
$ cd tmp
$ git clone -b "rpi-4.3.y" --single-branch https://github.com/raspberrypi/linux.git
```
I just tested this on a non-watchdog Pi and `oom-killer` said (among others):
```
Mar 21 18:05:50 rbups kernel: Out of memory: Kill process 1451 (git) score 849 or sacrifice child
Mar 21 18:05:50 rbups kernel: Killed process 1451 (git) total-vm:471580kB, anon-rss:418604kB, file-rss:176kB
```
The Pi with watchdog activated just reboots.

**Comment by diederikdehaas** [comment]
With what Pi version and how much RAM does it have?

**Comment by goranche** [comment]
> causes crashes

what's a "crash" in this context?

**Comment by Mausy5043** [comment]
Both are Pi1B+; 482MB total memory (GPU = 16MB)
Crash is defined here as: uncleanly rebooting where I expect the behaviour to be that `oom-killer` jumps in and saves the day.

The reboot looks more like a power-cycle than a clean `shutdown -r`. The logs abruptly stop.

**Comment by goranche** [comment]
uhm... I think you misunderstand what the watchdog does here... :thought_balloon: 

it resets the chip... that's it, there is no clean shutdown, there is nothing to save the day...

**Comment by Mausy5043** [comment]
Well, then I don't like the `watchdog`. `oom-killer` is much nicer. :wink:

**Comment by goranche** [comment]
the two have nothing in common, really, they solve completely different issues in different ways

oom-killer will not save you if the kernel hangs (panics)... it will not make sure a reboot will actually happen, even if something goes wrong, ... the hardware watchdog, however, can do those things

**Comment by Mausy5043** [comment]
Yes, but in this particular case it would have been nice if the watchdog would have waited a bit longer.
I don't know. Tired... need... sleep... :zzz:

**Comment by diederikdehaas** [comment]
> This log-entry bothers me
https://github.com/debian-pi/raspbian-ua-netinst/issues/298#issuecomment-167235672
IOW: _I found that the BCM2708 watchdog has a maximum 15 second timeout._

File: scripts/etc/init.d/rcS [hunk 1]
```diff
@@ -39,6 +39,7 @@ rootfstype=ext4
 final_action=reboot
 hardware_versions=detect
 hwrng_support=1
+enable_watchdog=1
```
Comment (Mausy5043) [inline scripts/etc/init.d/rcS]: I'd prefer this to read `enable_watchdog=0`.

Comment (diederikdehaas) [inline scripts/etc/init.d/rcS]: Me too

**Comment by goranche** [comment]
> the BCM2708 watchdog has a maximum 15 second timeout

:unamused: sheesh... I'll change systemds timeout to 14 seconds then...

File: scripts/etc/init.d/rcS (hunk 1 above)
Comment (goranche) [inline scripts/etc/init.d/rcS]: agree... will change

**Comment by goranche** [comment]
> to 14 seconds

why 14? excellent question, I've asked that myself as well... :innocent:

**Pull request merged**